### PR TITLE
Parallelize render, pre-fetch expression data in Image Pipeline (SCP-4542)

### PR DESCRIPTION
**Pull request by eweitz**
This speeds up how long it would take after upload to instantly view gene expression scatter plots.

[Previously](https://github.com/broadinstitute/single_cell_portal_core/pull/1571), the proof of concept rendered only four genes.  It did so serially and fetched expression data from SCP API.  

### Parallelization
Now, multiple instances of Puppeteer are launched in parallel.  For a machine with 4 CPU cores; this speeds up Image Pipeline by 3x.  One CPU core is left unused, to accommodate non-Image-Pipeline tasks.  This opens the door to reducing runtime by e.g. 30-60x by using many-CPU machines types from GCP.

### Get expression data from files
Requests for expressed data are now also intercepted.  By default, these frequent requests to SCP API from Puppeteer often fail due to opaque errors like `net::ERR_SSL_PROTOCOL_ERROR`.  So Image Pipeline does a direct `fetch` for the relevant SCP API before searching for a gene through the UI (which interestingly avoids those errors).  Then, when Puppeteer runs, it intercepts the frontend's request for expression from SCP API, and seamlessly replaces it with a file system read for pre-fetched expression data.

Importantly, this shows we can re-route SCP API expression data requests to use files.  So we could likely prepare all files in a preceding [scatter data cache job](https://docs.google.com/document/d/1-mhtoWrg3RHoDjWrHGv2h-KtlnnYGPg5-lxlgRyWaK0/edit#heading=h.5b86j42k5u0d), then use them in Image Pipeline to warm the image cache without notably burdening the Rails app server, and dramatically faster.  Image Pipeline would read expression data files from a local disk or bucket.  That's much safer and faster than flooding SCP API with requests that hit an uncached remote database undergirding production.

For a 1.3M cell study, this version of Image Pipeline can produce around 167 images in 26 minutes using 3 CPUs.  Using pre-generated scatter plot data would speed it up.  Handling timeouts and "Target closed" errors would improve robustness.   That potential follow-up work is tracked in SCP-4562 and SCP-4563.

To test:
0. This assumes you're set up per [previous PR](https://github.com/broadinstitute/single_cell_portal_core/pull/1571)
1. Start your local SCP
2. Get an accession for a public study where gene expression plots are visualizable
3. Run `export NODE_TLS_REJECT_UNAUTHORIZED=0; time node expression-scatter-plots.js --accession="<your accession>"`
4. Let it run for 2 minutes
5. Confirm output files exist in `output` directory

This satisfies SCP-4542.

### CODE REVIEW

File: image-pipeline/expression-scatter-plots.js [hunk 1]
```diff
@@ -20,87 +21,280 @@ const options = {
...
 
-// Make `images` directory if absent
-access('images', async err => {
-  if (err) {
-    await mkdir('images')
-  }
-})
+// Candidates for CLI argument
+// CPU count on Intel i7 is 1/2 of reported, due to hyperthreading
+const numCPUs = os.cpus().length / 2 - 1
```
Comment (bistline) [inline image-pipeline/expression-scatter-plots.js:26]: Non-blocking: shouldn't this also take into account newer m1/m2 chips?  Or is the calculation the same?

**Review by bistline — APPROVED** [review]
Looks good.  Very interested to see how much this will eventually speed up using the pre-render data artifacts.

File: image-pipeline/expression-scatter-plots.js (hunk 1 above)
Comment (eweitz) [reply image-pipeline/expression-scatter-plots.js:26]: Good point!  Running on Emily's M1 Pro reports: `Number of CPUs to be used on this client: 4`, so I think this is OK for now.  Ultimately, I imagine we'll drop the `/ 2` for the default value and add a CLI argument `--num-cpu` to override that.

**Review by ehanna4 — APPROVED** [review]
Unable to run this locally, paired with Eric and successfully ran against Staging to see the changes. We noted that my IP is IPv6 rather than IPv4 and I also am on an M1 so likely at least one of these factors contributed to my inability to get it working locally but don't think this should block this PR and can be investigated further in a future ticket. But this looks good and was cool to see working well on Staging 👍

**Review by jlchang — APPROVED** [review]
Tried to do a functional review and re-discovered that my laptop is only dual-core so:
`Number of CPUs to be used on this client: 1`
Can confirm that both image and json files were generated in the expected directory structure but can't comment on the parallelism.

**Pull request merged**